llvm/test/Transforms/LoopVectorize/AArch64/reduction-recurrence-costs-sve.ll

```diff
@@ -41,105 +140,117 @@ define i32 @chained_recurrences(i32 %x, i64 %y, ptr %src.1, i32 %z, ptr %src.2)
-; VSCALEFORTUNING2-NEXT:    [[WIDE_MASKED_GATHER10:%.*]] = call <vscale x 4 x i32> @llvm.masked.gather.nxv4i32.nxv4p0(<vscale x 4 x ptr> [[TMP44]], i32 4, <vscale x 4 x i1> splat (i1 true), <vscale x 4 x i32> poison)
-; VSCALEFORTUNING2-NEXT:    [[TMP45:%.*]] = or <vscale x 4 x i32> [[WIDE_MASKED_GATHER]], [[VEC_PHI]]
-; VSCALEFORTUNING2-NEXT:    [[TMP46:%.*]] = or <vscale x 4 x i32> [[WIDE_MASKED_GATHER8]], [[VEC_PHI5]]
-; VSCALEFORTUNING2-NEXT:    [[TMP47]] = or <vscale x 4 x i32> [[TMP45]], [[WIDE_MASKED_GATHER9]]
-; VSCALEFORTUNING2-NEXT:    [[TMP48]] = or <vscale x 4 x i32> [[TMP46]], [[WIDE_MASKED_GATHER10]]
-; VSCALEFORTUNING2-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], [[TMP4]]
+; VSCALEFORTUNING2-NEXT:    [[BROADCAST_SPLATINSERT5:%.*]] = insertelement <4 x i32> poison, i32 [[TMP24]], i64 0
+; VSCALEFORTUNING2-NEXT:    [[BROADCAST_SPLAT6]] = shufflevector <4 x i32> [[BROADCAST_SPLATINSERT5]], <4 x i32> poison, <4 x i32> zeroinitializer
+; VSCALEFORTUNING2-NEXT:    [[TMP19:%.*]] = shufflevector <4 x i32> [[VECTOR_RECUR]], <4 x i32> [[BROADCAST_SPLAT6]], <4 x i32> <i32 3, i32 4, i32 5, i32 6>
+; VSCALEFORTUNING2-NEXT:    [[TMP20]] = shufflevector <4 x i32> [[BROADCAST_SPLAT6]], <4 x i32> [[BROADCAST_SPLAT6]], <4 x i32> <i32 3, i32 4, i32 5, i32 6>
+; VSCALEFORTUNING2-NEXT:    [[TMP21:%.*]] = shufflevector <4 x i32> [[VECTOR_RECUR3]], <4 x i32> [[TMP19]], <4 x i32> <i32 3, i32 4, i32 5, i32 6>
+; VSCALEFORTUNING2-NEXT:    [[TMP22:%.*]] = shufflevector <4 x i32> [[TMP19]], <4 x i32> [[TMP20]], <4 x i32> <i32 3, i32 4, i32 5, i32 6>
+; VSCALEFORTUNING2-NEXT:    [[TMP23:%.*]] = extractelement <4 x i32> [[TMP21]], i32 0
+; VSCALEFORTUNING2-NEXT:    [[TMP69:%.*]] = or i32 [[TMP23]], [[X]]
+; VSCALEFORTUNING2-NEXT:    [[TMP25:%.*]] = extractelement <4 x i32> [[TMP22]], i32 0
+; VSCALEFORTUNING2-NEXT:    [[TMP26:%.*]] = or i32 [[TMP25]], [[X]]
+; VSCALEFORTUNING2-NEXT:    [[TMP27:%.*]] = shl i32 [[TMP69]], 1
+; VSCALEFORTUNING2-NEXT:    [[TMP28:%.*]] = shl i32 [[TMP26]], 1
+; VSCALEFORTUNING2-NEXT:    [[TMP29:%.*]] = or i32 [[TMP27]], 2
+; VSCALEFORTUNING2-NEXT:    [[TMP30:%.*]] = or i32 [[TMP28]], 2
+; VSCALEFORTUNING2-NEXT:    [[TMP31:%.*]] = or i32 [[TMP5]], [[TMP29]]
+; VSCALEFORTUNING2-NEXT:    [[TMP32:%.*]] = or i32 [[TMP5]], [[TMP30]]
+; VSCALEFORTUNING2-NEXT:    [[TMP33:%.*]] = or i32 [[TMP31]], [[X]]
+; VSCALEFORTUNING2-NEXT:    [[TMP34:%.*]] = or i32 [[TMP32]], [[X]]
+; VSCALEFORTUNING2-NEXT:    [[TMP35:%.*]] = load i32, ptr [[TMP11]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP36:%.*]] = load i32, ptr [[TMP13]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP37:%.*]] = load i32, ptr [[TMP15]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP38:%.*]] = load i32, ptr [[TMP17]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP39:%.*]] = insertelement <4 x i32> poison, i32 [[TMP35]], i32 0
```

Comment:
Do you know why we've vectorised this loop differently now? At first glance the code looks worse than before with all the inserts and extracts.

Reply:
The pointer operand of the GEP is an argument, which is an SCEVUnknown: this means that various instructions could be considered uniform in the chain and narrowed to scalar operations; I think the net result of vscale versus no-vscale vectorization was a cost-model decision? I'm not sure if the vectorization is worse overall, since many binops are operating on scalars?